src/pool/cli.c

```diff
@@ -1454,6 +1422,452 @@ dc_pool_query(tse_task_t *task)
+
+	n = pool_map_find_nodes(arg->mra_pool->dp_map, PO_COMP_ID_ALL, &nodes);
+	/* There must be at least one rank. */
+	D_ASSERTF(n > 0, "%d\n", n);
+
+	if (arg->mra_i == -1) {
+		/* Let i be a random integer in [0, n). */
+		i = ((double)rand() / RAND_MAX) * n;
+		if (i == n)
+			i = 0;
+	} else {
+		/* Continue the round robin. */
+		i = arg->mra_i;
+	}
+
+	/* Find next UPIN rank via a round robin from i. */
+	for (j = 0; j < n; j++) {
+		k = (i + j) % n;
+
+		if (nodes[k].do_comp.co_status == PO_COMP_ST_UPIN)
+			break;
+	}
+	/* There must be at least one UPIN rank. */
+	D_ASSERT(j < n);
+	D_ASSERT(k != -1);
+
+	arg->mra_i = k + 1;
+
+	return nodes[k].do_comp.co_rank;
+}
+
+static int
+create_map_refresh_rpc(struct dc_pool *pool, unsigned int map_version,
+		       crt_context_t ctx, crt_group_t *group, d_rank_t rank,
+		       crt_rpc_t **rpc, struct pool_buf **map_buf)
+{
+	crt_endpoint_t			ep;
+	crt_rpc_t		       *c;
+	struct pool_tgt_query_map_in   *in;
+	struct pool_buf		       *b;
+	int				rc;
+
+	ep.ep_grp = group;
+	ep.ep_rank = rank;
+
+	rc = pool_req_create(ctx, &ep, POOL_TGT_QUERY_MAP, &c);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to create POOL_TGT_QUERY_MAP: %d\n",
+			DP_UUID(pool->dp_pool), rc);
+		return rc;
+	}
+
+	in = crt_req_get(c);
+	uuid_copy(in->tmi_op.pi_uuid, pool->dp_pool);
+	uuid_copy(in->tmi_op.pi_hdl, pool->dp_pool_hdl);
+	in->tmi_map_version = map_version;
+
+	rc = map_bulk_create(ctx, &in->tmi_map_bulk, &b,
+			     pool_buf_nr(pool->dp_map_sz));
+	if (rc != 0) {
+		crt_req_decref(c);
+		return rc;
+	}
+
+	*rpc = c;
+	*map_buf = b;
+	return 0;
+}
+
+static void
+destroy_map_refresh_rpc(crt_rpc_t *rpc, struct pool_buf *map_buf)
+{
+	struct pool_tgt_query_map_in *in = crt_req_get(rpc);
+
+	map_bulk_destroy(in->tmi_map_bulk, map_buf);
+	crt_req_decref(rpc);
+}
+
+struct map_refresh_cb_arg {
+	crt_rpc_t	       *mrc_rpc;
+	struct pool_buf	       *mrc_map_buf;
+};
+
+static int
+map_refresh_cb(tse_task_t *task, void *varg)
+{
+	struct map_refresh_cb_arg      *cb_arg = varg;
+	struct map_refresh_arg	       *arg;
+	struct dc_pool		       *pool;
+	struct pool_tgt_query_map_in   *in = crt_req_get(cb_arg->mrc_rpc);
+	struct pool_tgt_query_map_out  *out = crt_reply_get(cb_arg->mrc_rpc);
+	struct pool_map		       *map;
+	bool				reinit = false;
+	int				rc = task->dt_result;
+
+	arg = tse_task_buf_embedded(task, sizeof(*arg));
+	pool = arg->mra_pool;
+
+	/*
+	 * If it turns out below that we do need to update the cached pool map,
+	 * then holding the lock while doing so will be okay, since we probably
+	 * do not want other threads to proceed with a known-stale pool anyway.
+	 * Otherwise, we will release the lock quickly.
+	 */
+	D_RWLOCK_WRLOCK(&pool->dp_map_lock);
+
+	D_DEBUG(DB_MD, DF_UUID": %p: crt: "DF_RC"\n", DP_UUID(pool->dp_pool),
+		task, DP_RC(rc));
+	if (daos_rpc_retryable_rc(rc)) {
+		reinit = true;
+		goto out;
+	} else if (rc != 0) {
+		goto out;
+	}
+
+	rc = out->tmo_op.po_rc;
+	if (rc == -DER_ALREADY) {
```

Comment:
it looks like only comparing the version would be enough here, i.e. does extra error code is needed?

Reply:
OK. Will compare the versions in the request and the reply instead.